os/binfmt/libelf/libelf_sections.c

```diff
@@ -208,64 +209,46 @@ void *elf_find_start_section_addr(struct binary_s *binp)
 
 void *elf_find_text_section_addr(int bin_idx)
 {
-	bin_addr_info_t *info;
-	info = (bin_addr_info_t *)sq_peek(&g_bin_addr_list);
-	while (info) {
-		if (info->bin_idx == bin_idx) {
-			return (void *)info->text_addr;
-		}
-		info = (FAR void *)sq_next((FAR sq_entry_t *)info);
+	if (bin_idx >= 0 && bin_idx <= CONFIG_NUM_APPS) {
+		return (void *)g_bin_addr_list[bin_idx].text_addr;
 	}
 	return NULL;
 }
 
 void elf_save_bin_section_addr(struct binary_s *bin)
 {
-	bin_addr_info_t *bin_info;
-	bin_info = (bin_addr_info_t *)kmm_malloc(sizeof(bin_addr_info_t));
-	if (bin_info != NULL) {
-		bin_info->bin_idx = bin->binary_idx;
-		bin_info->text_addr = bin->sections[BIN_TEXT];
-		bin_info->text_size = bin->sizes[BIN_TEXT];
+	uint8_t bin_idx = bin->binary_idx;
+
+	g_bin_addr_list[bin_idx].text_addr = bin->sections[BIN_TEXT];
+	g_bin_addr_list[bin_idx].text_size = bin->sizes[BIN_TEXT];
 #ifdef CONFIG_SAVE_BIN_SECTION_ADDR
-		binfo("[%s] text_addr : %x\n", bin->bin_name, bin_info->text_addr);
+	binfo("[%s] text_addr : %x\n", bin->bin_name, g_bin_addr_list[bin_idx].text_addr);
 #ifdef CONFIG_OPTIMIZE_APP_RELOAD_TIME
-		bin_info->rodata_addr = bin->sections[BIN_RO];
-		bin_info->data_addr = bin->sections[BIN_DATA];
-		bin_info->bss_addr = bin->sections[BIN_BSS];
+	g_bin_addr_list[bin_idx].rodata_addr = bin->sections[BIN_RO];
+	g_bin_addr_list[bin_idx].data_addr = bin->sections[BIN_DATA];
+	g_bin_addr_list[bin_idx].bss_addr = bin->sections[BIN_BSS];
 
-		binfo("   rodata_addr : %x\n", bin_info->rodata_addr);
-		binfo("   data_addr   : %x\n", bin_info->data_addr);
-		binfo("   bss_addr    : %x\n", bin_info->bss_addr);
+	binfo("   rodata_addr : %x\n", g_bin_addr_list[bin_idx].rodata_addr);
+	binfo("   data_addr   : %x\n", g_bin_addr_list[bin_idx].data_addr);
+	binfo("   bss_addr    : %x\n", g_bin_addr_list[bin_idx].bss_addr);
 #endif
 #endif
-		sq_addlast((sq_entry_t *)bin_info, &g_bin_addr_list);
-	} else {
-		berr("ERROR : Failed to allocate for bin symbol.\n");
-	}
 }
 
 void elf_delete_bin_section_addr(struct binary_s *bin)
```

Comment:
In this case, we don't need to pass the whole `bin`.
It is better to pass the `bin->binary_idx` only.
Please check this.